compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java

```diff
@@ -279,8 +286,25 @@ BLangStatementExpression desugar(BLangQueryExpr queryExpr, SymbolEnv env,
             BType refType = Types.getReferredType(queryExpr.getBType());
             if (TypeTags.isXMLTypeTag(refType.tag) || (refType.tag == TypeTags.UNION
                     && ((BUnionType) refType).getMemberTypes().stream().allMatch(memType ->
-                                    TypeTags.isXMLTypeTag(Types.getReferredType(memType).tag)))) {
-                result = getStreamFunctionVariableRef(queryBlock, QUERY_TO_XML_FUNCTION, Lists.of(streamRef), pos);
+                    TypeTags.isXMLTypeTag(Types.getReferredType(memType).tag == TypeTags.INTERSECTION ?
+                            ((BIntersectionType) Types.getReferredType(memType)).effectiveType.tag :
+                            Types.getReferredType(memType).tag)))) {
+                BLangLiteral isReadonly = new BLangLiteral(false, symTable.booleanType);
+                if (refType.tag == TypeTags.XML_TEXT || // The xml:Text type is inherently immutable.
```

Comment:
It's not required to specifically check for XML:Text here, since it's inherently immutable its readonly flag is on. We can just do the `isFlagOn` check